src/lib/json-rpc-router.ts

```diff
@@ -0,0 +1,37 @@
+import type { Request } from 'express';
+import type { Readable } from 'readable-stream';
+import type { JsonRpcRequest, JsonRpcResponse } from './json-rpc.js';
+
+type HttpRequest = {
+  contentType: 'application/octet-stream' | 'multipart/form-data';
+  request: Request;
+};
+
+export type RequestContext = {
+  transport?: 'http' | 'ws';
+} & (HttpRequest | { [K in keyof HttpRequest]?: never; });
+
+export type HandlerResponse = {
+  jsonRpcResponse: JsonRpcResponse,
+  dataStream?: Readable
+}
+
+export type JsonRpcHandler = (JsonRpcRequest: JsonRpcRequest, context: RequestContext) => Promise<HandlerResponse>
+
+export class JsonRpcRouter {
+  private methodHandlers: { [method: string]: JsonRpcHandler };
+
+  constructor() {
+    this.methodHandlers = {};
+  }
+
+  on(methodName: string, handler: JsonRpcHandler) {
+    this.methodHandlers[methodName] = handler;
+  }
+
+  async handle(rpcRequest: JsonRpcRequest, context: RequestContext = {}) {
+    const handler = this.methodHandlers[rpcRequest.method];
+
+    return await handler(rpcRequest, context);
```

Comment:
what if `handler` doesnt exist?

Reply:
it'd bork with `handler is not defined`. implying that `on` was called with `handler` as undefined which should be caught with tests

Reply:
either that or that `on` was never called with the relevant `method`

i only mention this cause the JSON-RPC spec has builtin errors to support "method not found", so perhaps we should use that here (and maybe "internal error" if the `method` does exist and the `handler` does not)?